dlgGeneralForGraphics, ucrAdditionalLayers, sdgPlots #1947

FrancoisJRenaud opened the pull request:
in progress...

FrancoisJRenaud commented:
Not quite convinced that it's the right moment to this. sdgPLots need to be rethought in my opinion. Maybe I m just used to work with sdgLayerOptions, but now that I m confrunted to the PlotOptions, I find the link very unnatural. One difference I see is that most of the settings on the Options are not editable on the main. But that starts to not be true anymore... We want to have coordinate flip on both, we have additional layers on general and options, ... 

This proposal still needs a lot of testing, probably quite a lot of editing. For the moment randomly disabled when no layers (following a wrong path initially). The biggest issue is what dataframe do we pass to facets ? If there is a layer, I m sure there is a dataframe in GlobalDataFrame. So left the condition in and gave that one through...

FrancoisJRenaud commented:
There are numerous bugs associated to the way sdgPlots is linked to other dialogues... There is a need for filling in information on the dialogue according to the data coming from main dlg. For the moment, sdgPlotOptions stores information about several dlg at a time. For example the ucrAdditionalLayers is really shared between all dialogues. I think somehting like a setup function similar to the setup function for LayerOptions would be good ? Things like the linking of the RSyntax with the sdgPlotRSyntax should not be performed in the initialise dialog of our main dialogues I think...

dannyparsons commented on instat/ucrAdditionalLayers.vb:
```diff
@@ -123,6 +123,8 @@ Public Class ucrAdditionalLayers
             lstLayerComplete.RemoveAt(lstLayers.SelectedIndices(0))
             lstLayers.Items.Remove(lstLayers.SelectedItems(0))
+
+            dlgGeneralForGraphics.TestOKEnabled()
```
What is this for?

FrancoisJRenaud replied:
This is for disabling ok when the last layer in the list is deleted. For the moment ok remains enabled. TestOkEnabled is also called in AddLayer...

dannyparsons replied:
But this control is not always on `dlgGeneralForGraphics` so this can't be how we do it.

FrancoisJRenaud replied:
yeah, true ... so I ll just add it after show plot options and delete it from here and there. I stupidly copied from other sub... There were a few other remains like that from a time where the control was only on dlgGeneralForGraphics, if this was ever the case

dannyparsons replied:
Yes, possibly. Something like LayersChanged.

FrancoisJRenaud replied:
oups chose a silly name: OkOrReturnInDanger... will change it to LayersChanged...

FrancoisJRenaud replied:
Mhm, disabling return is not usually done right. I guess it would be better to have a cancel button on sdg if we want to introduce disabling return. I m just thinking that the message explaining that the user needs to have all layers completed before clicking ok when he has created layers on options is not very nice ... But then here there is no need to desable return when the number of layers is zero as there is the main layer in general... will correct this

dannyparsons replied:
No we don't disable return. Not having a cancel is for the same reason we don't have a reset on sub dialogs, these are all part of general discussions on sub dialogs. For now, I'm happy if OK is just disabled on the main dialog if layers are incomplete.

FrancoisJRenaud commented:
So here are a few more changes, but the amount of work needed on the sdgPlots and links is rather huge and requires rewriting.... So this won't be done on this pull request. I don't think any bug was added from my testing, but certainly not much was solved... Things work as well as possible with generalForGraphics I think... also see  #2274

stevekogo commented on instat/sdgPlots.vb:
```diff
@@ -46,6 +46,13 @@ Public Class sdgPlots
         autoTranslate(Me)
     End Sub
 
+    Public Sub DisableLayersTab()
```
I don't think the sub routines  DisableLayersTab() and EnableLayersTab() are necessary since you can enable and disable layers with a single line command

FrancoisJRenaud replied:
Yup, I have the feeling that htings are a bit clearer like that, but then all this is going to be edited soon anyway so I don't have strong feelings...

FrancoisJRenaud commented:
testing works fine. Solves #2267

FrancoisJRenaud commented:
So in order to help reading this pull request: 

added options on general for graphics
corrected a few bugs in ucr additional layers in the choice of bIgnoreGlobalAes' value method and position of different sdg when adding layer via sdgPlots
raise event NumberOfLayersChanged in ucrAdditionalLayers when number of layers change (delete, add) and created a testoksub in general for graphics handling that event.
made a few modifications in sdgplots to accomodate options on generalforGraphics
also added the setup of ucrAdditionalLayers.clsGgplotFunction/RSyntax inside the functions that modify these on sdgPLots, as otherwise the links with the dlg calling for sdgPlots is lost. Solves  #2267

I know this is way too much for one pull request. These were just small changes identified during testing added day after day. Then the pull request was not merged (the reviewing got delayed here) but I needed the changes anyway for continuing testing. Little by little end up with this and just realised now I did it again. I hope the modifications are clear enough now. Sorry for the confusion... Testing is going fine. I promise I won't add anything else until merged.

dannyparsons commented:
@africanmathsinitiative/instat for testing a pull request like this which edits some core code - I suggest you discuss with @FrancoisJRenaud to understand what changes he's made, like he's written above, and then test out what he claims to have done on the branch on his computer (or you could clone the branch on your machien). I don't mind if you don't look as much at the code, but instead comment that you've tested the branch and it does/doesn't do what the author claims.

FrancoisJRenaud commented:
Yes I think that would work well. Thanks @dannyparsons ! We did that before for other pull requests. Will keep the habit.

maxwellfundi commented:
From the explanation from the above comment, we have gone through this pull request and the tested it this morning with @FrancoisJRenaud  and seems to work well as explained. 

Just one point to note,  the aesthetics doe not clear when we delete the layer. This might be solved by another pull request as this is an old branch and i think this code is now ready for @dannyparsons to review and merge.

Thanks.

FrancoisJRenaud commented:
>  the aesthetics doe not clear when we delete the layer. 

A little precision: when deleting all Layers, the global Aestethics are still in the GgplotFunction. I don't know if that's actually a bug. If the user deletes all layers one by one, instead of clicking reset, he might wish to indeed keep options and global aes he has set before ? Can also try to clear GlobalAes when there are no Layers anymore but not here... I need this branch to be merged to move on with the other pull request

FrancoisJRenaud commented:
@dannyparsons, really sorry to press a little, but do you think you will be able to merge soon ? The sdgLayerOptions is waiting as well, would like to finalise it when this one is merged and have both out of the way as I am really struggling with testing. Need to change branches all the time when I get a bug make sure it's a new one and not one I ve fixed on another pull request etc...

lilyclements commented:
So, before this is merged (and sorry Francois) I managed to crash it.. It might be an issue with sdgLayerOptions, however! I know there are several issues in that area which will be sorted eventually.

So, if I open lets say the scatterplot dialog, put in a single variable in the first selector, and then another variable in the "X Variable" selector. Then I go onto "Options" - "Additional Layer" and add in a new layer. Then I go back to the main dialog, and decide that I want a multiple variable now so change the first receiver to be a multiple variable, click "OK" and we get an error:

---------------------------
Error running R command(s)
---------------------------
External component has thrown an exception.

The error occurred in attempting to run the following R command(s):

border_dataframe_temp <- InstatDataObject$get_data_frame(data_name="border_dataframe", stack_data=TRUE, id.vars=c("no","runs"), measure.vars="no")

border_dataframe_temp <- InstatDataObject$get_data_frame(data_name="border_dataframe")

ggplot(data=border_dataframe_temp, mapping=aes(y=value, x=runs, colour=variable)) + geom_point() + xlab(label="") + ylab(label="") + geom_boxplot(mapping=aes(y=no), data=border_dataframe_temp)
---------------------------
OK   
---------------------------

FrancoisJRenaud commented:
@lilyclements  This is why we disabled options on most dialogues when these are in multiple variables case. Just forgot to do it for a few: scatter and rug, can do it in next pull request...

FrancoisJRenaud commented:
I think

lilyclements commented:
Okay, got it. Just thought I'd flag it up :)

FrancoisJRenaud commented:
Yup, good to see that these are tested intensely ! If that's the only bug you found I m quite happy ^^ I don't pretend to have solved everything. Just pretend to not make things worse... Please don't test the other pulls until I ve updated them with this one and gone through myself... There are a few things I need to check before peer review. But they need to be checked with the changes here...

FrancoisJRenaud commented:
Thanks !
